File: sdk/typescript/src/policy.ts
/**
 * CHAKRAVYUH OS — `POST /v1/policy/evaluate`
 *
 * Evaluates a custom or named security policy against a provided payload.
 * Supports dry-run mode for testing policies without side effects.
 */

import type {
  PolicyRequest,
  PolicyResponse,
  ProtectInput,
} from "./models";
import type { ChakravyuhHttpClient } from "./client";

/** Options for the `evaluatePolicy` method. */
export interface EvaluatePolicyOptions {
  /** If true, evaluates without creating audit records or modifying state. */
  dryRun?: boolean;
  /** Idempotency key. */
  idempotencyKey?: string;
  /** Correlation ID for distributed tracing. */
  correlationId?: string;
}

/**
 * Evaluate a security policy against a payload.
 *
 * Use this to:
 * - Test policy changes before deployment (dryRun: true)
 * - Evaluate complex custom policies
 * - Debug which rules match a given input
 * - Integrate policy decisions into external workflows
 *
 * @example
 * ```ts
 * const result = await ck.evaluatePolicy({
 *   policyId: "pol_custom_sql_injection",
 *   payload: {
 *     type: "api_request",
 *     content: "SELECT * FROM users WHERE id = 1 OR 1=1",
 *   },
 *   dryRun: true,
 * });
 * console.log(result.decision); // "deny"
 * ```
 */
export async function evaluatePolicy(
  client: ChakravyuhHttpClient,
  params: { policyId: string; payload: ProtectInput; dryRun?: boolean },
  options?: { idempotencyKey?: string; correlationId?: string },
): Promise<PolicyResponse> {
  const request: PolicyRequest = {
    policy_id: params.policyId,
    request_payload: params.payload,
    dry_run: params.dryRun,
  };

  return client.post<PolicyResponse>("/v1/policy/evaluate", request, options);
}

File: sdk/rust/src/policy.rs
/// Policy evaluate endpoint implementation.

use crate::client::HttpClient;
use crate::errors::ChakravyuhError;
use crate::models::{PolicyRequest, PolicyResponse, WithHeaders};

/// Optional parameters for the policy evaluate call.
#[derive(Debug, Clone, Default)]
pub struct PolicyOptions {
    /// Idempotency key for exactly-once processing.
    pub idempotency_key: Option<String>,
    /// Correlation ID for distributed tracing.
    pub correlation_id: Option<String>,
}

impl PolicyOptions {
    /// Create new empty options.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the idempotency key.
    pub fn idempotency_key(mut self, key: impl Into<String>) -> Self {
        self.idempotency_key = Some(key.into());
        self
    }

    /// Set the correlation ID.
    pub fn correlation_id(mut self, id: impl Into<String>) -> Self {
        self.correlation_id = Some(id.into());
        self
    }
}

/// Execute a policy evaluation request.
pub(crate) async fn execute_policy(
    client: &HttpClient,
    request: PolicyRequest,
    options: Option<PolicyOptions>,
) -> Result<WithHeaders<PolicyResponse>, ChakravyuhError> {
    let opts = options.unwrap_or_default();
    client
        .post_with_headers(
            "/v1/policy/evaluate",
            &request,
            opts.idempotency_key.as_deref(),
            opts.correlation_id.as_deref(),
        )
        .await
}